DOCS 5913 - Added Advanced Mule Concepts article. (#864)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/advanced-mule-concepts.adoc

= Advanced Mule Concepts
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: advanced, mule, concepts, runtime, engine

> This section covers some advanced Mule topics that will be useful once you are familiar with the basics of Mule application development and Mule runtime engine concepts.

Reviewer: GLOBALLY reword to avoid saying "this section" or "this topic" or any other self-referencing phrases.
We also want to avoid referring to "previous" or "next" topics. The goal is to make each topic self-supporting and independent, for eventual topic reuse (hahahahahahaha).

Reviewer: Just one way of doing this:
```suggestion
After you become familiar with the basics of Mule application development and Mule runtime engine (Mule) concepts, you can focus on more advanced features, such as the execution engine and third-party software upon which Mule depends.
```

> Learn about the xref:execution-engine.adoc[Execution Engine] to understand how the Mule runtime engine processes the operations in a Mule application’s flow, and understand how to change the thread pools configuration to optimize performance.

Reviewer 2: The phrase `change the thread pools configuration to optimize performance` is contradictory with the self-tuning approach that is already part of our message. I would rephrase it to just `change the thread pools configuration if needed`

> Learn about the xref:execution-engine.adoc[Execution Engine] to understand how the Mule runtime engine processes the operations in a Mule application’s flow, and understand how to change the thread pools if needed. Mule runtime engine uses execution metadata provided by operations in a flow to perform auto-tuning.

Reviewer: ```suggestion
Learn about the xref:execution-engine.adoc[execution engine] to understand how Mule runtime engine processes the operations in a Mule application flow, and to understand how to change the thread pools if needed. Mule runtime engine uses execution metadata provided by operations in a flow to perform autotuning.
```

> The xref:about-classloading-isolation.adoc[Classloading Isolation] mechanism implemented in Mule 4 addresses potential issues caused by the Java Virtual Machine when conflicting versions of the same JAR files exist in the same location.

Reviewer 2: `caused by the Java Virtual Machine when conflicting versions of the same JAR files exist in the same location.` -> `caused by conflicting versions of the same library are visible from the same location.`

> The xref:about-classloading-isolation.adoc[Classloading Isolation] mechanism implemented in Mule 4 addresses potential issues caused when conflicting versions of the same library are visible in the same location.

Reviewer 2: the original motivation of classloading isolation is to make upgrades easier by not allowing a component to expose its internals which may change, and components to interact between themselves through a well defined, complete and as-minimal-as-possible api.

> The xref:about-classloading-isolation.adoc[Classloading Isolation] mechanism implemented in Mule 4 is designed to make upgrades easier by preventing components from exposing their internal resources. All components interact with each other through their defined APIs, meaning that only exposed resources can be accessed by other components. Classloading isolation ensures that a component's internal content is protected and can be changed anytime without affecting other components.

Reviewer: ```suggestion
The xref:about-classloading-isolation.adoc[classloading isolation] mechanism implemented in Mule 4 is designed to make upgrades easier by preventing components from exposing their internal resources. All components interact with each other through their defined APIs so that only exposed resources can be accessed by other components. Classloading isolation ensures that a component's internal content is protected and can be changed any time, without affecting other components.
```

> Mule 4 supports xref:distributed-file-polling.adoc[Distributed File Polling], which allows connectors to read files from certain directories as they are created, and it also makes it possible to poll files in all cluster nodes.

Reviewer: ```suggestion
Mule 4 supports xref:distributed-file-polling.adoc[distributed file polling], which allows connectors to read files from certain directories as they are created. Distributed file polling also makes it possible to poll files in all cluster nodes.
```

You can check the xref:mule-4-api-javadoc.adoc[Javadocs] for the Mule Kernel, the Community Edition of the Mule runtime engine, to see definitions for classes, fields and their data types.

> MuleSoft also provides a list of xref:third-party-software-in-mule.adoc[Third-party software] that is included as part of the source code, examples, default services or as dependencies.

Reviewer 2: `provides` doesn't sound good, since the libs bundled in the runtime are for internal use of the runtime, while `provide` would be interpreted that it is something that you may use from an app.

> MuleSoft also provides a list of xref:third-party-software-in-mule.adoc[Third-party software] that is included as part of the source code, examples, default services or as dependencies.

Reviewer 2: I don't understand what the last part of the sentence means.

> You can check the xref:mule-4-api-javadoc.adoc[Javadocs] for the Mule Kernel, the Community Edition of the Mule runtime engine, to see definitions for classes, fields and their data types.

Reviewer: ```suggestion
You can refer to the xref:mule-4-api-javadoc.adoc[Javadocs] for Mule Kernel (the community edition of the Mule runtime engine) to learn definitions for classes, fields, and data types.
```

> You can see a list of xref:third-party-software-in-mule.adoc[Third-party software] that is included as part of Mule runtime engine as internal dependencies.

Reviewer: ```suggestion
You can see a list of xref:third-party-software-in-mule.adoc[third-party software] that is included as part of Mule runtime engine as internal dependencies.
```
